File: Data/src/com/dlgdev/pokemon/database/DatabaseDefinition.java
package com.dlgdev.pokemon.database;

import java.util.Collections;
import java.util.List;

public class DatabaseDefinition {

	public static class Pokemon {
		public static final String TABLE_NAME = "pokemon";
		public static final String POKEMON_DEX_NUMBER = "pokemon_id";
		public static final String FORM_ID = "form_id";
		public static final String HEIGHT = "height";
		public static final String WEIGHT = "weight";
		public static final String COLOR = "color";
		public static final String MALE_RATIO = "";

		public static final String primaryKey = String.format("%s,%s", POKEMON_DEX_NUMBER, FORM_ID);

	}

	public static class Evolutions {
		public static final String TABLE_NAME = "evolutions";
		public static final String FROM = "pokemon_id";
		public static final String TO_DEX = "dex";
		public static final String TO_FORM = "form";

		public static final String primaryKey = String.format("%s,%s,%s", FROM, TO_DEX, TO_FORM);
		public static final List<String> indexes = Collections.emptyList();
		static {
			indexes.add(FROM);
			indexes.add(TO_DEX);
			indexes.add(TO_FORM);
		}
	}

	public static class PokemonNames {
		public static final String TABLE_NAME = "pokemonname";
		public static final String POKEMON_FORM_ID = "form_id";
		public static final String POKEMON_NAME = "pokemon_name";
		public static final String LANGUAGE = "language";

		public static final String primaryKey = String.format("%s,%s", POKEMON_FORM_ID, LANGUAGE);
		public static final List<String> indexes = Collections.emptyList();

		static {
			indexes.add(POKEMON_NAME);
		}
	}

	public static class PokemonBaseStats {
		public static final String TABLE_NAME = "pokemonbasestats";
		public static final String POKEMON_FORM_ID = "form_id";
		public static final String GEN = "gen";
		public static final String HP = "hp";
		public static final String ATK = "atk";
		public static final String DEF = "def";
		public static final String SPATK = "spatk";
		public static final String SPDEF = "spdef";
		public static final String SPEED = "speed";

		public static final String primaryKey = String.format("%s,%s", POKEMON_FORM_ID, GEN);
	}

	public static class PokemonTypes {
		public static final String TABLE_NAME = "pokemontypes";
		public static final String POKEMON_FORM_ID = "form_id";
		public static final String GEN = "gen";
		public static final String TYPE_1 = "type1";
		public static final String TYPE_2 = "type2";

		public static final String primaryKey = String.format("%s,%s", POKEMON_FORM_ID, GEN);
		public static final List<String> indexes = Collections.emptyList();

		static {
			indexes.add(TYPE_1);
			indexes.add(TYPE_2);
		}
	}

	public static class PokemonAbilities {
		public static final String TABLE_NAME = "pokemonabilities";
		public static final String POKEMON_FORM_ID = "form_id";
		public static final String GEN = "gen";
		public static final String ABILITY_1 = "ability1";
		public static final String ABILITY_2 = "ability2";
		public static final String ABILITY_HIDDEN = "hiddenAbility";

		public static final String primaryKey = String.format("%s,%s", POKEMON_FORM_ID, GEN);
		public static final List<String> indexes = Collections.emptyList();

		static {
			indexes.add(ABILITY_1);
			indexes.add(ABILITY_2);
			indexes.add(ABILITY_HIDDEN);
		}
	}

	public static class Abilities {
		public static final String TABLE_NAME = "abilities";
		public static final String ABILITY_ID = "ability_id";
		public static final String NAME = "ability_name";
		public static final String LANGUAGE = "language";

		public static final String primaryKey = String.format("%s,%s", ABILITY_ID, LANGUAGE);
	}

	public static class PokemonEggGroups {
		public static final String TABLE_NAME = "pokemonegggroups";
		public static final String POKEMON_ID = "pokemon_id";
		public static final String EGG_GROUP = "egg_group";

		public static final String primaryKey = String.format("%s,%s", POKEMON_ID, EGG_GROUP);
		public static final List<String> indexes = Collections.emptyList();

		static {
			indexes.add(EGG_GROUP);
		}
	}
}

File: database/src/main/kotlin/com/dlgdev/pokemon/database/DatabaseDefinition.kt
package com.dlgdev.pokemon.database

class DatabaseDefinition {

    object Pokemon {
        val TABLE_NAME = "pokemon"
        val ID = "pokemon_id"
        val DEX_NUMBER = "dex_number"
        val FORM = "form"

        val uniqueKey = "$DEX_NUMBER,$FORM"
    }

    object PokemonData {
        val TABLE_NAME = "pokemondata"
        val POKEMON_ID = "pokemon_id"
        val HEIGHT = "base_height"
        val WEIGHT = "base_weight"
        val COLOR = "color"
        val MALE_RATIO = "maleratio"
    }

    object Evolutions {
        val TABLE_NAME = "evolutions"
        val GAME = "game_id"
        val FROM = "from_id"
        val TO = "to_id"
        val EVOLUTION_METHOD = "evolution_method"
        val EVOLUTION_POINT = "evolution_point"
        val IS_MEGA = "is_mega"
    }

    object PokemonNames {
        val TABLE_NAME = "pokemonnames"
        val POKEMON_ID = "pokemon_id"
        val POKEMON_NAME = "name"
        val FORM_NAME = "form_name"
        val LANGUAGE = "language"
    }

    object PokemonBaseStats {
        val TABLE_NAME = "pokemonbasestats"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val HP = "hp"
        val ATK = "atk"
        val DEF = "def"
        val SPATK = "spatk"
        val SPDEF = "spdef"
        val SPEED = "speed"
    }

    object PokemonTypes {
        val TABLE_NAME = "pokemontypes"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val TYPE_1 = "type_1"
        val TYPE_2 = "type_2"

    }

    object PokemonAbilities {
        val TABLE_NAME = "pokemonabilities"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val ABILITY_1 = "ability_1"
        val ABILITY_2 = "ability_2"
        val ABILITY_HIDDEN = "ability_hidden"
    }

    object Abilities {
        val TABLE_NAME = "abilities"
        val ABILITY_ID = "ability_id"
        val GAME = "game_id"
        val NAME = "ability_name"
        val DESCRIPTION = "description"
        val LANGUAGE = "language"
    }

    object PokemonEggGroups {
        val TABLE_NAME = "pokemonegggroups"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val EGG_GROUP = "egg_group"
    }

    object Games {
        val TABLE_NAME = "games"
        val ID = "game_id"
        val NAME = "name"
        val GEN = "gen_id"
    }

    object Generations {
        val TABLE_NAME = "generations"
        val ID = "gen_id"
        val NAME = "name"
        val REPRESENTATION = "representation"
    }
    object Attacks {
        val TABLE_NAME = "attacks"
        val ID = "attack_id"
        val GEN = "gen_id"
        val BASE_POWER = "basepower"
        val ACCURACY = "accuracy"
        val ATTACK_TYPE = "attacktype"
        val MODE = "attackmode" //Special, Physical, Other
    }

    object AttackNames {
        val TABLE_NAME = "attacknames"
        val ATTACK_ID = "attack_id"
        val ATTACK_NAME = "attack_name"
        val DESCRIPTION = "attack_description"
        val LANGUAGE = "language"
    }
    
    object AttacksWithEffects {
        val TABLE_NAME = "attackswitheffects"
        val ATTACK_ID = "attack_id"
        val EFFECT_ID = "effect_id"
        val GEN = "gen_id"
        val CHANCE = "chance"
    }
    
    object AttackEffects {
        val TABLE_NAME = "attackeffects"
        val ID = "effect_id"
        val GAME = "game_id"
        val EFFECT_TYPE = "effect_type"
    }

    object PokemonLevelAttacks {
        val TABLE_NAME = "pokemonattackslevel"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val LEVEL = "level"
        val ATTACK_ID = "attack_id"
    }

    object PokemonTMAttacks {
        val TABLE_NAME = "pokemonattackstm"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val TM_NUMBER = "tm_number"
        val ATTACK_ID = "attack_id"
    }

    object PokemonTutorAttacks {
        val TABLE_NAME = "pokemonattackstutor"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val ATTACK_ID = "attack_id"
    }

    object PokemonEventAttacks {
        val TABLE_NAME = "pokemonattacksevent"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val ATTACK_ID = "attack_id"
    }

    object CatchRates {
        val TABLE_NAME = "catchrates"
        val POKEMON_ID = "pokemon_id"
        val GAME = "game_id"
        val CATCH_RATE = "catch_rate"
    }
}
